crates/env/src/lib.rs

```diff
@@ -99,3 +99,24 @@ pub use self::{
         NoChainExtension,
     },
 };
+pub use ink_prelude;
+
+/// Appends a formatted string to the `debug_message` buffer which will be:
+///  - Returned to the caller when the contract is invoked via RPC (*not* via an extrinsic)
+///  - Logged as a `debug!` message on the substrate node, which will be printed to the node
+///    console's `stdout` when the log level is set to `debug`.
+#[macro_export]
+macro_rules! debug_print {
+    ($($arg:tt)*) => ($crate::debug_message(&$crate::ink_prelude::format!($($arg)*)));
+}
+
+/// Appends a formatted string to the `debug_message` buffer, as per [`debug_print`] but
+/// with a newline appended.
+#[macro_export]
+macro_rules! debug_println {
+    () => ($crate::print!("\n"));
+    ($($arg:tt)*) => ({
+        $crate::debug_print!($($arg)*);
+        $crate::debug_print!("\n");
```

Comment:
I'd prefer using a single call to `debug_print` if that's easily possible.

Reply:
Simple alternative is something like
```
let msg = $crate::ink_prelude::format!($($arg)*);
$crate::debug_message(&$crate::ink_prelude::format!("{}\n", &msg))
```
Or could be done inline. 

Anything more involved to avoid allocations might be overkill since this should never be executed as part of a transaction.

What do you think?

Reply:
Out of interest Rust's `println!` macro looks like it interns the string in a compiler builtin with a `\n` appended: https://github.com/rust-lang/rust/blob/673d0db5e393e9c64897005b470bfeb6d5aec61b/compiler/rustc_builtin_macros/src/format.rs#L944

Reply:
This is a debug API so I do not care too much about extraneous allocations. However, it would be really confusing to the user if a single call to `debug_message!` always ends up in 2 actual calls to the engine. Imo we definitely need to fix this. I'd accept the allocation solution for now but would mark it with a comment that it is possible to optimize it further using our static 16kb buffer under the hood instead of allocating. If you are into it you can also toy around with this idea.

Reply:
I agree. Calls into the supervisor should be kept to a minimum. They carry a non trivial overhead even when debugging is disabled (on-chain execution).